pulse-rs/src/stream.rs

```diff
@@ -0,0 +1,413 @@
+// Copyright © 2017 Mozilla Foundation
+//
+// This program is made available under an ISC-style license.  See the
+// accompanying file LICENSE for details.
+
+use ::*;
+use context;
+use ffi;
+use operation;
+use std::ffi::CStr;
+use std::mem;
+use std::os::raw::{c_int, c_void};
+use std::ptr;
+use util::*;
+
+#[derive(Debug)]
+pub struct Stream(*mut ffi::pa_stream);
+
+impl Stream {
+    pub fn new<'a, CM>(c: &Context, name: &::std::ffi::CStr, ss: &SampleSpec, map: CM) -> Self
+        where CM: Into<Option<&'a ChannelMap>>
+    {
+        unsafe {
+            stream::from_raw_ptr(ffi::pa_stream_new(c.raw_mut(),
+                                                    name.as_ptr(),
+                                                    ss as *const _,
+                                                    to_ptr(map.into())))
+        }
+    }
+
+    #[doc(hidden)]
+    pub fn raw_mut(&self) -> &mut ffi::pa_stream {
+        unsafe { &mut *self.0 }
+    }
+
+    pub fn is_null(&self) -> bool {
+        self.0.is_null()
+    }
+
+    pub fn get_state(&self) -> StreamState {
+        StreamState::try_from(unsafe {
+            ffi::pa_stream_get_state(self.raw_mut())
+        }).expect("pa_stream_get_state returned invalid StreamState")
```

Comment:
We should turn most of the `expect`s in this code into Result returns in the future.